File: cxx-mnp-plasmon/src/mnp_plasmon.cpp
#include "mnp_plasmon.hpp"
#include <iostream>
#include <iomanip>
#include <algorithm>

namespace mnp {

// Material database definition
const std::vector<Material> MnpPlasmon::MATERIALS = {
    {"Au", 3.106, 0.0132, 8.90},
    {"Ag", 3.810, 0.0048, 3.91},
    {"Al", 14.83, 0.098, 1.24}
};

std::vector<Material> MnpPlasmon::material_list() {
    return MATERIALS;
}

bool MnpPlasmon::material_exists(const std::string& name) {
    return std::any_of(MATERIALS.begin(), MATERIALS.end(),
        [&name](const Material& m) { return m.name == name; });
}

Material MnpPlasmon::material_get(const std::string& name) {
    auto it = std::find_if(MATERIALS.begin(), MATERIALS.end(),
        [&name](const Material& m) { return m.name == name; });
    
    if (it != MATERIALS.end()) {
        return *it;
    }
    throw std::invalid_argument("Material not found: " + name);
}

double MnpPlasmon::wavelength_to_energy(double wavelength_nm) {
    return HC_EV_NM / wavelength_nm;
}

complex MnpPlasmon::drude_epsilon(const std::string& material, double wavelength_nm) {
    if (!material_exists(material)) {
        throw std::invalid_argument("Unknown material: " + material);
    }
    
    Material mat = material_get(material);
    double omega = wavelength_to_energy(wavelength_nm);
    
    // omega^2
    double omega_sq = omega * omega;
    
    // i*gamma*omega
    complex damping(0.0, omega * mat.gamma);
    
    // omega_p^2 / (omega^2 + i*gamma*omega)
    complex denominator = omega_sq + damping;
    complex drude_term = (mat.omega_p * mat.omega_p) / denominator;
    
    return complex(mat.eps_inf, 0.0) - drude_term;
}

complex MnpPlasmon::constant_epsilon(double n_squared) {
    return complex(n_squared, 0.0);
}

double MnpPlasmon::complex_magnitude(complex c) {
    return std::abs(c);
}

double MnpPlasmon::complex_magnitude_sq(complex c) {
    return std::norm(c);
}

complex MnpPlasmon::rayleigh_polarizability(
    double radius_nm,
    complex eps_particle,
    complex eps_medium
) {
    // 4*pi*a^3
    double volume_factor = 4.0 * PI * radius_nm * radius_nm * radius_nm;
    
    // (eps_p - eps_m) / (eps_p + 2*eps_m)
    complex numerator = eps_particle - eps_medium;
    complex denominator = eps_particle + 2.0 * eps_medium;
    
    if (complex_magnitude(denominator) < 1e-10) {
        return complex(0.0, 0.0);
    }
    
    return volume_factor * numerator / denominator;
}

double MnpPlasmon::wave_vector(double wavelength_nm, double n_medium) {
    return 2.0 * PI * n_medium / wavelength_nm;
}

CrossSections MnpPlasmon::rayleigh_cross_sections(
    double wavelength_nm,
    double radius_nm,
    complex eps_particle,
    complex eps_medium
) {
    CrossSections sections;
    
    complex alpha = rayleigh_polarizability(radius_nm, eps_particle, eps_medium);
    double k = wave_vector(wavelength_nm, eps_medium.real());
    
    double k_pow4 = k * k * k * k;
    double alpha_mag_sq = complex_magnitude_sq(alpha);
    double alpha_im = alpha.imag();
    
    sections.c_ext = k * alpha_im;
    sections.c_sca = (k_pow4 / (6.0 * PI)) * alpha_mag_sq;
    sections.c_abs = sections.c_ext - sections.c_sca;
    
    // Ensure non-negative absorption
    if (sections.c_abs < 0.0) {
        sections.c_abs = 0.0;
    }
    
    return sections;
}

SphereResponse MnpPlasmon::simulate_sphere_response(
    const std::string& material,
    double wavelength_nm,
    double radius_nm,
    double medium_refractive_index
) {
    SphereResponse response;
    
    response.wavelength_nm = wavelength_nm;
    response.radius_nm = radius_nm;
    response.medium_refractive_index = medium_refractive_index;
    
    // Calculate permittivity
    response.epsilon_particle = drude_epsilon(material, wavelength_nm);
    complex eps_medium = constant_epsilon(medium_refractive_index * medium_refractive_index);
    
    // Calculate polarizability
    response.polarizability = rayleigh_polarizability(
        radius_nm,
        response.epsilon_particle,
        eps_medium
    );
    
    // Calculate cross-sections
    CrossSections cs = rayleigh_cross_sections(
        wavelength_nm,
        radius_nm,
        response.epsilon_particle,
        eps_medium
    );
    
    response.c_ext = cs.c_ext;
    response.c_sca = cs.c_sca;
    response.c_abs = cs.c_abs;
    
    return response;
}

void MnpPlasmon::print_response(const SphereResponse& response) {
    std::cout << "=== MNP Plasmon Response ===\n";
    std::cout << std::fixed << std::setprecision(4);
    std::cout << "Wavelength: " << response.wavelength_nm << " nm\n";
    std::cout << "Radius: " << response.radius_nm << " nm\n";
    std::cout << "Medium n: " << response.medium_refractive_index << "\n";
    std::cout << "\nDielectric Function:\n";
    std::cout << "  ε = " << response.epsilon_particle.real() << " + " 
              << response.epsilon_particle.imag() << "i\n";
    std::cout << "\nPolarizability:\n";
    std::cout << std::scientific << std::setprecision(6);
    std::cout << "  α = " << response.polarizability.real() << " + " 
              << response.polarizability.imag() << "i nm³\n";
    std::cout << "\nCross-Sections:\n";
    std::cout << "  C_ext = " << response.c_ext << " nm²\n";
    std::cout << "  C_sca = " << response.c_sca << " nm²\n";
    std::cout << "  C_abs = " << response.c_abs << " nm²\n";
    std::cout << "=============================\n";
}

}  // namespace mnp

File: csharp-mnp-plasmon/MnpPlasmon/MnpPlasmon.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Numerics;

namespace MnpPlasmon;

/// <summary>
/// Material structure for Drude model parameters
/// </summary>
public record Material(
    string Name,
    double OmegaP,      // Plasma frequency (eV)
    double Gamma,       // Damping coefficient (eV)
    double EpsInf       // High-frequency dielectric constant
);

/// <summary>
/// Output structure for sphere response
/// </summary>
public record SphereResponse(
    double WavelengthNm,
    double RadiusNm,
    double MediumRefractiveIndex,
    Complex EpsParticle,
    Complex Polarizability,
    double CExt,        // Extinction cross-section (nm²)
    double CSca,        // Scattering cross-section (nm²)
    double CAbs         // Absorption cross-section (nm²)
);

/// <summary>
/// Cross-section result
/// </summary>
public record CrossSections(
    double CExt,
    double CSca,
    double CAbs
);

/// <summary>
/// MnpPlasmon - C# implementation for optical response of metallic nanoparticles
/// </summary>
public static class MnpPlasmon
{
    // Constants
    private const double HcEvNm = 1239.84197;  // hc in eV·nm
    private const double PI = Math.PI;

    // Material database
    private static readonly Material[] Materials =
    {
        new("Au", 3.106, 0.0132, 8.90),
        new("Ag", 3.810, 0.0048, 3.91),
        new("Al", 14.83, 0.098, 1.24)
    };

    /// <summary>
    /// Get list of available materials
    /// </summary>
    public static IReadOnlyList<Material> MaterialList() => Array.AsReadOnly(Materials);

    /// <summary>
    /// Check if material exists in database
    /// </summary>
    public static bool MaterialExists(string name) =>
        Materials.Any(m => m.Name == name);

    /// <summary>
    /// Get material from database
    /// </summary>
    public static Material GetMaterial(string name) =>
        Materials.FirstOrDefault(m => m.Name == name)
            ?? throw new ArgumentException($"Material not found: {name}");

    /// <summary>
    /// Convert wavelength (nm) to energy (eV)
    /// </summary>
    private static double WavelengthToEnergy(double wavelengthNm) =>
        HcEvNm / wavelengthNm;

    /// <summary>
    /// Drude model: epsilon(omega) = eps_inf - omega_p^2 / (omega^2 + i*gamma*omega)
    /// </summary>
    public static Complex DrudeEpsilon(string material, double wavelengthNm)
    {
        var mat = GetMaterial(material);
        double omega = WavelengthToEnergy(wavelengthNm);

        // omega^2
        double omegaSq = omega * omega;

        // i*gamma*omega
        var damping = new Complex(0, omega * mat.Gamma);

        // omega_p^2 / (omega^2 + i*gamma*omega)
        var denominator = new Complex(omegaSq, 0) + damping;
        var druideTerm = (mat.OmegaP * mat.OmegaP) / denominator;

        return new Complex(mat.EpsInf, 0) - druideTerm;
    }

    /// <summary>
    /// Constant permittivity (non-absorbing dielectric)
    /// epsilon = n^2
    /// </summary>
    public static Complex ConstantEpsilon(double nSquared) =>
        new(nSquared, 0);

    /// <summary>
    /// Rayleigh polarizability
    /// alpha = 4*pi*a^3 * (eps_p - eps_m) / (eps_p + 2*eps_m)
    /// </summary>
    public static Complex RayleighPolarizability(
        double radiusNm,
        Complex epsParticle,
        Complex epsMedium)
    {
        // 4*pi*a^3
        double volumeFactor = 4.0 * PI * radiusNm * radiusNm * radiusNm;

        // (eps_p - eps_m) / (eps_p + 2*eps_m)
        var numerator = epsParticle - epsMedium;
        var denominator = epsParticle + 2.0 * epsMedium;

        if (denominator.Magnitude < 1e-10)
            return Complex.Zero;

        return volumeFactor * numerator / denominator;
    }

    /// <summary>
    /// Wave vector k = 2*pi*n_medium / wavelength_nm
    /// </summary>
    private static double WaveVector(double wavelengthNm, double nMedium) =>
        2.0 * PI * nMedium / wavelengthNm;

    /// <summary>
    /// Rayleigh cross-sections
    /// C_ext = k * Im(alpha)
    /// C_sca = |k|^4 / (6*pi) * |alpha|^2
    /// C_abs = C_ext - C_sca
    /// </summary>
    public static CrossSections RayleighCrossSections(
        double wavelengthNm,
        double radiusNm,
        Complex epsParticle,
        Complex epsMedium)
    {
        var alpha = RayleighPolarizability(radiusNm, epsParticle, epsMedium);
        double k = WaveVector(wavelengthNm, epsMedium.Real);

        double kPow4 = k * k * k * k;
        double alphaMagSq = alpha.Magnitude * alpha.Magnitude;
        double alphaIm = alpha.Imaginary;

        double cExt = k * alphaIm;
        double cSca = (kPow4 / (6.0 * PI)) * alphaMagSq;
        double cAbs = cExt - cSca;

        // Ensure non-negative absorption
        if (cAbs < 0.0)
            cAbs = 0.0;

        return new CrossSections(cExt, cSca, cAbs);
    }

    /// <summary>
    /// High-level API: simulate sphere optical response
    /// </summary>
    public static SphereResponse SimulateSphereResponse(
        string material,
        double wavelengthNm,
        double radiusNm,
        double mediumRefractiveIndex)
    {
        // Calculate permittivity
        var epsParticle = DrudeEpsilon(material, wavelengthNm);
        var epsMedium = ConstantEpsilon(mediumRefractiveIndex * mediumRefractiveIndex);

        // Calculate polarizability
        var polarizability = RayleighPolarizability(radiusNm, epsParticle, epsMedium);

        // Calculate cross-sections
        var cs = RayleighCrossSections(wavelengthNm, radiusNm, epsParticle, epsMedium);

        return new SphereResponse(
            wavelengthNm,
            radiusNm,
            mediumRefractiveIndex,
            epsParticle,
            polarizability,
            cs.CExt,
            cs.CSca,
            cs.CAbs
        );
    }

    /// <summary>
    /// Pretty-print response structure
    /// </summary>
    public static void PrintResponse(SphereResponse response)
    {
        Console.WriteLine("=== MNP Plasmon Response ===");
        Console.WriteLine($"Wavelength: {response.WavelengthNm:F2} nm");
        Console.WriteLine($"Radius: {response.RadiusNm:F2} nm");
        Console.WriteLine($"Medium n: {response.MediumRefractiveIndex:F4}");
        Console.WriteLine();
        Console.WriteLine("Dielectric Function:");
        Console.WriteLine($"  ε = {response.EpsParticle.Real:F6} + {response.EpsParticle.Imaginary:F6}i");
        Console.WriteLine();
        Console.WriteLine("Polarizability:");
        Console.WriteLine($"  α = {response.Polarizability.Real:E6} + {response.Polarizability.Imaginary:E6}i nm³");
        Console.WriteLine();
        Console.WriteLine("Cross-Sections:");
        Console.WriteLine($"  C_ext = {response.CExt:E6} nm²");
        Console.WriteLine($"  C_sca = {response.CSca:E6} nm²");
        Console.WriteLine($"  C_abs = {response.CAbs:E6} nm²");
        Console.WriteLine("=============================");
    }
}
